sam_view.c

```diff
@@ -560,211 +560,425 @@ static const char *copied_tags[] = { "RG", "BC", "QT", NULL };
+{
+    bam_hdr_destroy(state->h);
+    check_sam_close(state->fp, opts->fn_input, "file", status);
+    if (state->fpse && !fclose(state->fpse)) { print_error_errno("Error closing singleton file \"%s\"", opts->fnse); return true; }
+    if (state->fpr[0] != stdout && fclose(state->fpr[0])) { print_error_errno("Error closing r0 file"); return true; }
+    if (state->fpr[1] != stdout && fclose(state->fpr[1])) { print_error_errno("Error closing r1 file \"%s\"", opts->fnr1); return true; }
+    if (state->fpr[2] != stdout && fclose(state->fpr[2])) { print_error_errno("Error closing r2 file \"%s\"", opts->fnr2); return true; }
+    free(state->fpr);
+    free(state);
+    return false;
+}
+
+static bool bam2fq_mainloop_singletontrack(bam2fq_state_t *state)
+{
+    // process a name collated BAM into fastq
+    bam1_t* b = bam_init1();
+    if (b == NULL) {
+        perror(NULL);
+        return true;
+    }
+    int64_t n_singletons = 0, n_reads = 0; // Statistics
+    kstring_t linebuf = { 0, 0, NULL }; // Buffer
+    char *previous_readname = NULL; // The readname of the read in linebuf
+    int previous_readpart = READ_UNKNOWN; // Whether the read in linebuf belongs to the forward, reverse, or unknown read
     kputsn("", 0, &linebuf);
 
-    while (sam_read1(fp, h, b) >= 0) {
-        if (b->core.flag&(BAM_FSECONDARY|BAM_FSUPPLEMENTARY)) continue; // skip secondary and supplementary alignments
+    // This loop writes reads as pairs, holding the unpaired read from the previous read in linebuf until either
+    // the read's mate is found or a new read name is encountered in which case we write it as a singleton.
+    while (sam_read1(state->fp, state->h, b) >= 0) {
+        if (b->core.flag&(BAM_FSECONDARY|BAM_FSUPPLEMENTARY) // skip secondary and supplementary alignments
+            || (b->core.flag&(state->flag_on)) != state->flag_on             // or reads indicated by filter flags
+            || (b->core.flag&(state->flag_off)) != 0) continue;
         ++n_reads;
 
-        int i;
-        int32_t qlen = b->core.l_qseq;
-        assert(qlen >= 0);
-        uint8_t* seq;
-        uint8_t* qual = bam_get_qual(b);
-        const uint8_t *oq = NULL;
-        if (use_oq) oq = bam_aux_get(b, "OQ");
-        bool has_qual = (qual[0] != 0xff || (use_oq && oq)); // test if there is quality
-
+        //////////////////////////////////////////////////////////////////
+        // First deal with the previous read, checking if it has a mate //
+        //////////////////////////////////////////////////////////////////
         // If there was a previous readname
-        if ( fpse && previous ) {
-            if (!strcmp(bam_get_qname(b), previous ) ) {
-                fputs(linebuf.s, stdout); // Write previous read
-                free(previous);
-                previous = NULL;
-            } else { // Doesn't match it's a singleton
+        if ( previous_readname ) {
```

Comment:
By necessity the singleton code needs to assume read name sorted data so it can check the previous read name to identify singles vs pairs.

For paired data, it'll produce output on position sorted files, but it may or may not be helpful depending on the downstream application.   We should add documentation explaining this.